### PR TITLE
Automatically install latest spark version

### PR DESCRIPTION
**Pull request by mathbunnyru**
## Issue ticket if applicable

Fix: https://github.com/jupyter/docker-stacks/issues/2063
Close: https://github.com/jupyter/docker-stacks/issues/1937

## Checklist (especially for first-time contributors)

- [ ] I have performed a self-review of my code
- [ ] If it is a core feature, I have added thorough tests
- [ ] I will try not to use force-push to make the review process easier for reviewers
- [ ] I have updated the documentation for significant changes

### CODE REVIEW

File: docs/using/specifics.md [hunk 1]
```diff
@@ -39,51 +39,10 @@ ipython profile create
...
-  - `openjdk_version`: The version of the OpenJDK (JRE headless) distribution (`17`).
-    - This version needs to match the version supported by the Spark distribution used above.
-    - See [Spark Overview](https://spark.apache.org/docs/latest/#downloading) and [Ubuntu packages](https://packages.ubuntu.com/search?keywords=openjdk).
-
-- Starting with _Spark >= 3.2_, the distribution file might contain the Scala version.
-
-For example, here is how to build a `pyspark-notebook` image with Spark `3.2.0`, Hadoop `3.2`, and OpenJDK `11`.
-
-```{warning}
-This recipe is not tested and might be broken.
```
Comment (mathbunnyru) [inline docs/using/specifics.md:61]: This recipe doesn't work right now and the easiest thing is to remove it at all.
If someone wants to use old spark versions, we have tags.

And python script is not that difficult to change (or if someone wants to add customization but with proper test, I will gladly merge such a PR).

**Comment by consideRatio** [comment]
I just wanted to say that I think the comments motivating changes made sense and seem to me like good calls! Thank you for your amazing work @mathbunnyru!!!

**Comment by mathbunnyru** [comment]
Thank you for the review @consideRatio, and your kind words.

I think I'm done with this PR 🙂
If someone thinks removing the recipe and spark customization is a bad idea, please let me know.

**Comment by bjornjorgensen** [comment]
You can build a `pyspark-notebook` image with a different `Spark` version by adjusting
the ['setup_spark.py'](https://github.com/jupyter/docker-stacks/blob/main/images/pyspark-notebook/setup_spark.py) script.

How are users going to do that?

**Comment by mathbunnyru** [comment]
> You can build a `pyspark-notebook` image with a different `Spark` version by adjusting
> 
> the ['setup_spark.py'](https://github.com/jupyter/docker-stacks/blob/main/images/pyspark-notebook/setup_spark.py) script.
> 
> 
> 
> How are users going to do that? 
> 
> 
> 
> 

If users want to build their own images, I assume they are able to change our source code. So they can hardcode the version in the script, fix example.
Not the best solution, but I doubt this feature was actively used.

**Comment by bjornjorgensen** [comment]
> If users want to build their own images, I assume they are able to change our source code. So they can hardcode the version in the script, fix example. Not the best solution, but I doubt this feature was actively used.

No, thats not right. Apache spark maintain a branch for 18 month so I know that users are on different branches.

**Comment by mathbunnyru** [comment]
Ok, @bjornjorgensen, I will add customisation back in a separate PR.

**Comment by mathbunnyru** [comment]
> Ok, @bjornjorgensen, I will add customisation back in a separate PR after merging this one.

Done right here.

File: docs/using/specifics.md [hunk 2]
```diff
@@ -65,14 +68,14 @@ This recipe is not tested and might be broken.
...
 # Build the image with different arguments
 docker build --rm --force-rm \
-    -t jupyter/pyspark-notebook:spark-3.2.0 ./images/pyspark-notebook \
+    -t my-spark-notebook ./images/pyspark-notebook \
+    --build-arg openjdk_version=11 \
     --build-arg spark_version=3.2.0 \
     --build-arg hadoop_version=3.2 \
-    --build-arg spark_checksum=707DDE035926A50B75E53FCA72CADA519F3239B14A96546911CB4916A58DCF69A1D2BFDD2C7DD5899324DBD82B6EEAB9797A7B4ABF86736FFCA4C26D0E0BF0EE \
-    --build-arg openjdk_version=11
+    --build-arg spark_download_url="https://archive.apache.org/dist/spark/"
```
Comment (mathbunnyru) [inline docs/using/specifics.md:75]: Works on my machine now.

**Comment by mathbunnyru** [comment]
These commands do work:

`make build/pyspark-notebook`

This one is in the docs:
```
docker build --rm --force-rm \
     -t my-pyspark-notebook ./images/pyspark-notebook \
     --build-arg openjdk_version=11 \
     --build-arg spark_version=3.2.0 \
     --build-arg hadoop_version=3.2 \
     --build-arg spark_download_url="https://archive.apache.org/dist/spark/"
```

This tests the Scala:
```
docker build --rm --force-rm \
    -t my-pyspark-notebook ./images/pyspark-notebook \
    --build-arg scala_version=2.13
```

The last one reuses the layer with jdk installation from default build now, which is also nice.

I am supper-happy with the work I've done here.
@consideRatio @bjornjorgensen please, take a look.

**Comment by bjornjorgensen** [comment]
These commands do work:

make build/pyspark-notebook

can you update the docs with this...

**Comment by bjornjorgensen** [comment]
are there anyway you can find the pandas version that are for the spark version? 
if not we must update it ones there are a new spark version out.

**Comment by mathbunnyru** [comment]
> These commands do work:
> 
> make build/pyspark-notebook
> 
> can you update the docs with this...

I do not want to mention Makefile in our docs - this Makefile is not tested at all (and is not used in any part of the building process). It is quite stable though and experienced users can still use it.

**Comment by bjornjorgensen** [comment]
> > These commands do work:
> > make build/pyspark-notebook
> > can you update the docs with this...
> 
> I do not want to mention Makefile in our docs - this Makefile is not tested at all (and is not used in any part of the building process). It is quite stable though and experienced users can still use it.

ok, but there should be some docs for how users can build the image without of using the latest branch of spark

**Comment by mathbunnyru** [comment]
> are there anyway you can find the pandas version that are for the spark version? if not we must update it ones there are a new spark version out.

Yes and no.
I thought about this.
The dependency is specified in a format `'pandas<=2.1.4'` (at least now). It means that old pinned version should probably still work when a new Spark is out (if Spark doesn't specify the lower bound one day).
I know almost nothing about Spark infrastructure - if there is a way to get this version without regex'ing some file in a branch in another repo, that would be nice.

And I'm certain I won't do it in this PR (it's simply too much), although I am ready to merge even a PR with regex'ing to make it 100% automatic.

**Comment by mathbunnyru** [comment]
> > > These commands do work:
> > > make build/pyspark-notebook
> > > can you update the docs with this...
> > 
> > 
> > I do not want to mention Makefile in our docs - this Makefile is not tested at all (and is not used in any part of the building process). It is quite stable though and experienced users can still use it.
> 
> ok, but there should be some docs for how users can build the image without of using the latest branch of spark

Well, take a look at the file `docs/using/specifics.md`, I changed it a little bit to make it work, but it has exactly this information.

**Comment by bjornjorgensen** [comment]
one more thing when we are her..
spark used to be installed at opt/spark https://github.com/apache/spark/blob/e4335206b59a9aafd8d3616d260d1b7834a72856/resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/python/Dockerfile#L36C28-L36C28

**Comment by mathbunnyru** [comment]
> one more thing when we are her.. spark used to be installed at opt/spark https://github.com/apache/spark/blob/e4335206b59a9aafd8d3616d260d1b7834a72856/resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/python/Dockerfile#L36C28-L36C28

I think this is an orthogonal feature.
Also, running the current `main` branch and my PR produces the same result.

I also think our users are used to having SPARK_HOME at `/usr/local/spark` symlink.
I haven't heard any complaints about this location, so I don't see any reason to change the Spark dir currently (and we don't want to break existing users).

**Comment by mathbunnyru** [comment]
Merging this one 🙂

**Comment by mathbunnyru** [comment]
@bjornjorgensen btw, `dev/infra/Dockerfile` didn't exist in before `v3.4.0`: https://github.com/apache/spark/tree/v3.3.4/dev

In the master branch, it has a slightly another format than even in `v3.5.0`.

master: `ARG BASIC_PIP_PKGS="numpy pyarrow>=14.0.0 six==1.16.0 pandas<=2.1.4 scipy unittest-xml-reporting plotly>=4.8 mlflow>=2.8.1 coverage matplotlib openpyxl memory-profiler>=0.61.0 scikit-learn>=1.3.2"`
v3.5.0: `RUN python3.9 -m pip install numpy pyarrow 'pandas<=2.0.3' scipy unittest-xml-reporting plotly>=4.8 'mlflow>=2.3.1' coverage matplotlib openpyxl 'memory-profiler==0.60.0' 'scikit-learn==1.1.*'`

So, if someone wants to cover all this to make it work for all Spark versions, I'm ready to merge it.
But I won't be doing it myself.

**Pull request merged**